### PR TITLE
feat: support AsyncAPI v2.4.0

### PR DESCRIPTION
**Pull request by owais34**
**Description**
AsyncAPI v2.4.0 readiness

Test Results :

> cross-env CI=true jest --coverage --testPathIgnorePatterns ./test/blackbox --testPathIgnorePatterns ./examples/TEMPLATE

Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
 PASS  test/models/Draft7Schema.spec.ts (8.655 s)
 PASS  test/models/CommonModel.spec.ts (10.963 s)
 PASS  test/processors/JsonSchemaInputProcessor.spec.ts (12.352 s)
 PASS  test/interpreter/unit/InterpretNot.spec.ts
 PASS  test/interpreter/unit/Intepreter.spec.ts
 PASS  test/generators/javascript/JavaScriptGenerator.spec.ts (17.143 s)
 PASS  test/interpreter/PostInterpreter.spec.ts (30.633 s)
 PASS  test/generators/java/JavaGenerator.spec.ts (30.667 s)
 PASS  test/generators/go/GoGenerator.spec.ts (22.059 s)
 PASS  test/generators/csharp/CSharpGenerator.spec.ts (17.221 s)
 PASS  test/generators/typescript/TypeScriptGenerator.spec.ts (30.779 s)
 PASS  test/generators/typescript/preset/MarshallingPreset.spec.ts (31.068 s)
 PASS  test/models/OpenAPIV3Schema.spec.ts
 PASS  test/processors/AsyncAPIInputProcessor.spec.ts
 PASS  test/models/SwaggerV2Schema.spec.ts
 PASS  test/generators/csharp/CSharpRenderer.spec.ts
 PASS  test/processors/InputProcessor.spec.ts
 PASS  test/generators/java/presets/CommonPreset.spec.ts
 PASS  test/utils/LoggingInterface.spec.ts
 PASS  test/generators/AbstractRenderer.spec.ts
 PASS  test/interpreter/unit/InterpretItems.spec.ts
 PASS  test/interpreter/unit/InterpretAllOf.spec.ts
 PASS  test/helpers/FormatHelpers.spec.ts
 PASS  test/generators/java/JavaRenderer.spec.ts
 PASS  test/interpreter/unit/Utils.spec.ts
 PASS  test/models/AsyncapiV2Schema.spec.ts
 PASS  test/generators/AbstractGenerator.spec.ts
 PASS  test/generators/typescript/TypeScriptRenderer.spec.ts
 PASS  test/interpreter/unit/InterpretAdditionalProperties.spec.ts
 PASS  test/processors/OpenAPIInputProcessor.spec.ts
 PASS  test/generators/typescript/preset/utils/ExampleFunctions.spec.ts
 PASS  test/interpreter/unit/InterpretAdditionalItems.spec.ts
 PASS  test/processors/SwaggerInputProcessor.spec.ts
 PASS  test/interpreter/unit/InterpretProperties.spec.ts
 PASS  test/generators/go/GoRenderer.spec.ts
 PASS  test/interpreter/unit/patternProperties.spec.ts
 PASS  test/helpers/NameHelpers.spec.ts
 PASS  test/generators/javascript/preset/utils/ExampleFunctions.spec.ts
 PASS  test/models/Draft4Schema.spec.ts
 PASS  test/models/Draft6Schema.spec.ts
 PASS  test/generators/javascript/JavaScriptFileGenerator.spec.ts
 PASS  test/generators/java/presets/DescriptionPreset.spec.ts
 PASS  test/generators/typescript/TypeScriptFileGenerator.spec.ts
 PASS  test/interpreter/unit/InterpretEnum.spec.ts
 PASS  test/interpreter/unit/InterpretDependencies.spec.ts
 PASS  test/interpreter/unit/InterpretConst.spec.ts
 PASS  test/models/OutputModel.spec.ts
 PASS  test/helpers/TypeHelpers.spec.ts
 PASS  test/generators/java/presets/ConstraintsPreset.spec.ts
 PASS  test/generators/csharp/CSharpFileGenerator.spec.ts
 PASS  test/generators/java/JavaFileGenerator.spec.ts
 PASS  test/generators/java/presets/JacksonPreset.spec.ts
 PASS  test/generators/javascript/JavaScriptRenderer.spec.ts
 PASS  test/helpers/FileHelpers.spec.ts
 PASS  test/generators/csharp/presets/CommonPreset.spec.ts
 PASS  test/generators/csharp/presets/JsonSerializerPreset.spec.ts
 PASS  test/generators/javascript/preset/MarshallingPreset.spec.ts
 PASS  test/generators/typescript/preset/DescriptionPreset.spec.ts
 PASS  test/models/RenderOutput.spec.ts
 PASS  test/generators/javascript/preset/ExamplePreset.spec.ts
 PASS  test/generators/typescript/preset/ExamplePreset.spec.ts
 PASS  examples/typescript-enum-type/index.spec.ts
 PASS  test/generators/typescript/renderers/EnumRenderer.spec.ts
 PASS  test/interpreter/Intepreter.spec.ts
 PASS  examples/generate-to-files/index.spec.ts
 PASS  examples/custom-logging/index.spec.ts
 PASS  examples/generate-go-enums/index.spec.ts
 PASS  examples/typescript-use-esm/index.spec.ts
 PASS  examples/javascript-use-esm/index.spec.ts
 PASS  examples/swagger2.0-from-object/index.spec.ts
 PASS  examples/typescript-use-cjs/index.spec.ts
 PASS  examples/openapi-from-object/index.spec.ts
 PASS  examples/csharp-generate-serializer/index.spec.ts
 PASS  examples/javascript-use-cjs/index.spec.ts
 PASS  examples/adapting-input-and-output/index.spec.ts
 PASS  examples/javascript-generate-marshalling/index.spec.ts
 PASS  examples/csharp-generate-equals-and-hashcode/index.spec.ts
 PASS  examples/typescript-generate-marshalling/index.spec.ts
 PASS  examples/java-generate-marshalling/index.spec.ts
 PASS  examples/java-generate-tostring/index.spec.ts
 PASS  examples/java-generate-hashcode/index.spec.ts
 PASS  examples/java-generate-equals/index.spec.ts
 PASS  examples/indentation-type-and-size/index.spec.ts
 PASS  examples/javascript-generate-example/index.spec.ts
 PASS  examples/typescript-generate-example/index.spec.ts
 PASS  examples/csharp-auto-implemented-properties/index.spec.ts
 PASS  examples/csharp-change-collection-type/index.spec.ts
 PASS  examples/java-generate-jackson-annotation/index.spec.ts
 PASS  examples/json-schema-draft7-from-object/index.spec.ts
 PASS  examples/java-change-collection-type/index.spec.ts
 PASS  examples/asyncapi-from-parser/index.spec.ts (5.11 s)
 PASS  examples/asyncapi-from-object/index.spec.ts
 PASS  examples/generate-javascript-models/index.spec.ts
 PASS  examples/include-custom-function/index.spec.ts
 PASS  examples/typescript-interface/index.spec.ts
 PASS  examples/generate-csharp-models/index.spec.ts
 PASS  examples/java-generate-javax-constraint-annotation/index.spec.ts
 PASS  examples/generate-typescript-models/index.spec.ts
 PASS  examples/typescript-generate-comments/index.spec.ts
 PASS  examples/generate-java-models/index.spec.ts
 PASS  examples/java-generate-javadoc/index.spec.ts
 PASS  examples/generate-go-models/index.spec.ts
 PASS  test/processors/TypeScriptInputProcessor.spec.ts (25.654 s)
 PASS  examples/java-from-typescript-type-with-options/index.spec.ts (22.731 s)
 PASS  examples/java-from-typescript-type/index.spec.ts (18.412 s)
-----------------------------------------|---------|----------|---------|---------|--------------------------------------
File                                     | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-----------------------------------------|---------|----------|---------|---------|--------------------------------------
All files                                |   96.94 |    86.15 |   96.72 |   97.37 | 
 src                                     |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators                          |     100 |    88.57 |     100 |     100 | 
  AbstractGenerator.ts                   |     100 |    85.71 |     100 |     100 | 110,124
  AbstractRenderer.ts                    |     100 |    90.47 |     100 |     100 | 46-47
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/csharp                   |   90.62 |     73.8 |   83.33 |   92.39 | 
  CSharpFileGenerator.ts                 |     100 |      100 |     100 |     100 | 
  CSharpGenerator.ts                     |   86.11 |    54.54 |     100 |   88.57 | 71-72,78-79
  CSharpPreset.ts                        |     100 |      100 |     100 |     100 | 
  CSharpRenderer.ts                      |   89.47 |    80.64 |   66.66 |   91.89 | 50,66-67
  Constants.ts                           |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/csharp/presets           |     100 |       88 |     100 |     100 | 
  CommonPreset.ts                        |     100 |    83.33 |     100 |     100 | 20,32-54,80
  JsonSerializerPreset.ts                |     100 |     92.3 |     100 |     100 | 44,149
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/csharp/renderers         |     100 |    88.15 |     100 |     100 | 
  ClassRenderer.ts                       |     100 |    89.47 |     100 |     100 | 21-35,61,87,95-99
  EnumRenderer.ts                        |     100 |    84.21 |     100 |     100 | 50,79-93
 src/generators/go                       |    85.6 |    67.07 |   86.95 |   88.13 | 
  Constants.ts                           |     100 |      100 |     100 |     100 | 
  GoFileGenerator.ts                     |   36.36 |      100 |       0 |      40 | 16-22
  GoGenerator.ts                         |      80 |    46.42 |     100 |   84.31 | 30,39-42,79-80,86-87
  GoPreset.ts                            |     100 |      100 |     100 |     100 | 
  GoRenderer.ts                          |     100 |    76.92 |     100 |     100 | 26,59,61-73,115
  index.ts                               |     100 |      100 |       0 |     100 | 
 src/generators/go/renderers             |     100 |      100 |     100 |     100 | 
  EnumRenderer.ts                        |     100 |      100 |     100 |     100 | 
  StructRenderer.ts                      |     100 |      100 |     100 |     100 | 
 src/generators/java                     |   97.47 |     79.8 |   95.45 |   97.41 | 
  Constants.ts                           |     100 |      100 |     100 |     100 | 
  JavaFileGenerator.ts                   |     100 |      100 |     100 |     100 | 
  JavaGenerator.ts                       |   94.44 |       60 |     100 |   94.44 | 48-49
  JavaPreset.ts                          |     100 |      100 |     100 |     100 | 
  JavaRenderer.ts                        |   98.36 |    84.52 |     100 |   98.33 | 69
  index.ts                               |     100 |      100 |       0 |     100 | 
 src/generators/java/presets             |   96.96 |    83.75 |     100 |   96.87 | 
  CommonPreset.ts                        |   97.29 |    80.76 |     100 |   97.14 | 136-137
  ConstraintsPreset.ts                   |   93.54 |       90 |     100 |   93.54 | 39,43
  DescriptioPreset.ts                    |     100 |    83.33 |     100 |     100 | 17                                   
  JacksonPreset.ts                       |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/java/renderers           |     100 |    90.38 |     100 |     100 | 
  ClassRenderer.ts                       |     100 |    91.89 |     100 |     100 | 19,40,73
  EnumRenderer.ts                        |     100 |    86.66 |     100 |     100 | 23-37
 src/generators/javascript               |     100 |    72.91 |   94.44 |     100 | 
  Constants.ts                           |     100 |      100 |     100 |     100 | 
  JavaScriptFileGenerator.ts             |     100 |      100 |     100 |     100 | 
  JavaScriptGenerator.ts                 |     100 |    84.61 |     100 |     100 | 48
  JavaScriptPreset.ts                    |     100 |      100 |     100 |     100 | 
  JavaScriptRenderer.ts                  |     100 |    66.66 |     100 |     100 | 32,34-58
  index.ts                               |     100 |      100 |       0 |     100 | 
 src/generators/javascript/presets       |     100 |    78.57 |     100 |     100 | 
  CommonPreset.ts                        |     100 |    78.57 |     100 |     100 | 20-27,105-112,167-198
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/javascript/presets/utils |   91.42 |    80.95 |     100 |   91.42 | 
  ExampleFunction.ts                     |   91.42 |    80.95 |     100 |   91.42 | 19,43,56
 src/generators/javascript/renderers     |     100 |    85.71 |     100 |     100 | 
  ClassRenderer.ts                       |     100 |    85.71 |     100 |     100 | 31,73
 src/generators/typescript               |    98.7 |    84.25 |   96.66 |   98.65 | 
  Constants.ts                           |     100 |      100 |     100 |     100 | 
  TypeScriptFileGenerator.ts             |     100 |      100 |     100 |     100 | 
  TypeScriptGenerator.ts                 |     100 |     90.9 |     100 |     100 | 53,70
  TypeScriptPreset.ts                    |     100 |      100 |     100 |     100 | 
  TypeScriptRenderer.ts                  |   97.05 |    80.28 |     100 |   96.87 | 110,169
  index.ts                               |     100 |      100 |       0 |     100 | 
 src/generators/typescript/presets       |   95.65 |     91.3 |   81.48 |   95.61 | 
  CommonPreset.ts                        |     100 |    85.71 |     100 |     100 | 27,112,167-198
  DescriptionPreset.ts                   |   86.66 |      100 |   71.42 |   86.66 | 54-59
  ExportKeywordPreset.ts                 |   57.14 |      100 |      40 |   57.14 | 26-36
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/generators/typescript/presets/utils |   94.59 |       88 |     100 |   94.59 | 
  ExampleFunction.ts                     |   94.59 |       88 |     100 |   94.59 | 53,69
 src/generators/typescript/renderers     |   99.13 |    78.57 |     100 |   99.12 | 
  ClassRenderer.ts                       |     100 |    81.81 |     100 |     100 | 30,71,95,107
  EnumRenderer.ts                        |   97.22 |    71.42 |     100 |   97.22 | 80
  InterfaceRenderer.ts                   |     100 |      100 |     100 |     100 | 
  TypeRenderer.ts                        |     100 |    83.33 |     100 |     100 | 26
 src/helpers                             |    96.9 |    91.86 |     100 |   97.64 | 
  FileHelpers.ts                         |     100 |      100 |     100 |     100 | 
  FormatHelpers.ts                       |   95.45 |    92.15 |     100 |   97.43 | 165
  NameHelpers.ts                         |   95.65 |    95.45 |     100 |   95.23 | 49
  PresetHelpers.ts                       |     100 |    33.33 |     100 |     100 | 16
  TypeHelpers.ts                         |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/interpreter                         |   99.56 |    92.26 |     100 |     100 | 
  InterpretAdditionalItems.ts            |     100 |      100 |     100 |     100 | 
  InterpretAdditionalProperties.ts       |     100 |      100 |     100 |     100 | 
  InterpretAllOf.ts                      |     100 |      100 |     100 |     100 | 
  InterpretConst.ts                      |     100 |      100 |     100 |     100 |                                      
  InterpretDependencies.ts               |     100 |      100 |     100 |     100 | 
  InterpretEnum.ts                       |     100 |      100 |     100 |     100 | 
  InterpretItems.ts                      |     100 |    83.33 |     100 |     100 | 28,32
  InterpretNot.ts                        |   94.44 |    73.33 |     100 |     100 | 14-15,24,28
  InterpretPatternProperties.ts          |     100 |      100 |     100 |     100 | 
  InterpretProperties.ts                 |     100 |      100 |     100 |     100 | 
  Interpreter.ts                         |     100 |    82.92 |     100 |     100 | 37,59,62-67,87,97,113
  PostInterpreter.ts                     |     100 |    95.23 |     100 |     100 | 25
  Utils.ts                               |     100 |      100 |     100 |     100 | 
 src/models                              |   99.31 |    97.43 |     100 |   99.51 | 
  AsyncapiV2Schema.ts                    |     100 |      100 |     100 |     100 | 
  CommonInputModel.ts                    |     100 |      100 |     100 |     100 | 
  CommonModel.ts                         |   98.61 |    95.97 |     100 |   99.02 | 31,40
  Draft4Schema.ts                        |     100 |      100 |     100 |     100 | 
  Draft6Schema.ts                        |     100 |      100 |     100 |     100 | 
  Draft7Schema.ts                        |     100 |      100 |     100 |     100 | 
  OpenapiV3Schema.ts                     |     100 |      100 |     100 |     100 | 
  OutputModel.ts                         |     100 |      100 |     100 |     100 | 
  Preset.ts                              |     100 |      100 |     100 |     100 | 
  RenderOutput.ts                        |     100 |      100 |     100 |     100 | 
  SwaggerV2Schema.ts                     |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/processors                          |    92.3 |    80.08 |   97.91 |    93.4 | 
  AbstractInputProcessor.ts              |     100 |      100 |     100 |     100 | 
  AsyncAPIInputProcessor.ts              |   84.78 |    80.26 |   88.88 |   85.88 | 73,79,82,85,88,91,94,100,103-106,123
  InputProcessor.ts                      |     100 |      100 |     100 |     100 | 
  JsonSchemaInputProcessor.ts            |      93 |    78.82 |     100 |   93.43 | 158,174,180,209,232,235,240,243,246 
  OpenAPIInputProcessor.ts               |   98.27 |       80 |     100 |     100 | 34,57,75-82
  SwaggerInputProcessor.ts               |     100 |    81.25 |     100 |     100 | 54-56,68
  TypeScriptInputProcessor.ts            |   83.33 |    78.57 |     100 |   87.17 | 47-50,56
  index.ts                               |     100 |      100 |     100 |     100 | 
 src/utils                               |     100 |      100 |     100 |     100 | 
  LoggingInterface.ts                    |     100 |      100 |     100 |     100 | 
  index.ts                               |     100 |      100 |     100 |     100 | 
-----------------------------------------|---------|----------|---------|---------|--------------------------------------

Test Suites: 105 passed, 105 total
Tests:       662 passed, 662 total
Snapshots:   116 passed, 116 total
Time:        70.13 s
Ran all test suites.
- ...
- ...
- ...

**Related issue(s)**

### CODE REVIEW

**Comment by owais34** [comment]
@jonaslagoni the streetlights-operation-security.yml file to be placed in which folder ?

**Comment by jonaslagoni** [comment]
> @jonaslagoni the streetlights-operation-security.yml file to be placed in which folder ?

In the 2.4 folder, as it uses the new operation security model 🙂

**Comment by jonaslagoni** [comment]
Do you mind converting it to JSON btw? Don't think the BlackBox test setup can handle YAML 😄

**Comment by owais34** [comment]
@jonaslagoni  Sure will do

**Comment by owais34** [comment]
@jonaslagoni done

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
@owais34 do you mind rebase with master and upgrade the parser to `1.15` that should automatically fix the blackbox tests that are failing 🙂

**Comment by owais34** [comment]
@jonaslagoni  I think that should do it hopefully , this was quite a learning experience  😁.

File: test/blackbox/docs/AsyncAPI-2_4/streetlights-operation-security.json [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            }
+          ],
+          "message": {
+            "$ref": "#/components/messages/lightMeasured"
+          },
+          "security": {
+            "streetlights_auth": [
+              "streetlights:write"
+            ]
+          }
```
Comment (jonaslagoni) [inline test/blackbox/docs/AsyncAPI-2_4/streetlights-operation-security.json:61]: The security keyword expects an array 🙂 Take a look here: https://github.com/asyncapi/spec/blob/65d58696f8c3ab6471aa48b12d878221b8d8109e/examples/streetlights-operation-security.yml#L64
```suggestion
          "security": [{
            "streetlights_auth": [
              "streetlights:write"
            ]
          }]
```

You need to apply this to the rest of the channels 🙂

**Comment by owais34** [comment]
@jonaslagoni  yesssss!!!!🤩

**Comment by jonaslagoni** [comment]
@all-contributors please add @owais34 for code, test, docs

**Pull request merged**